Lib/_pyrepl/unix_console.py

```
@@ -390,7 +390,15 @@ def restore(self):
             os.write(self.output_fd, b"\033[?7h")
 
         if hasattr(self, "old_sigwinch"):
-            signal.signal(signal.SIGWINCH, self.old_sigwinch)
+            # Only restore signal handler if we're in the main thread
+            # signal.signal() only works in the main thread of the main interpreter
+            try:
+                signal.signal(signal.SIGWINCH, self.old_sigwinch)
+            except ValueError:
```

We shouldn't just silence exceptions. Is there a way to reliably make this work without a wide try-catch? maybe ValueError is raised due to some other factors.

---

I am not sure, since it can only happen in mac
will try to dig it tonight

---

Well, you can check if you're on a darwin platform, and also check if we're in the same interpreter's thread. But maybe the issue is instead in the `asyncio` REPL. In the future, I would advise investigating a bit more (and avoid solutions that simply silence the issue with a try-except) before jumping on a PR.

---

copy that, thank you.

---

Oh nevermind. There is a similar code block above that does exactly this. So the question now is "is it correct to silently suppress ValueError"? if so, keep your code, if not, let's think about a better solution.

---

When defining `self.old_sigwinch`, we try-catch ValueError, but that's likely in order to check if the signal is actually supported by the platform, and I don't think it's to suppress an error that occurred because of the thread we're in.

---

> Well, you can check if you're on a darwin platform, and also check if we're in the same interpreter's thread. But maybe the issue is instead in the `asyncio` REPL. In the future, I would advise investigating a bit more (and avoid solutions that simply silence the issue with a try-except) before jumping on a PR.

and FYI the root casuse I already add it to the comments, not without any investigating

---

> When defining self.old_sigwinch, we try-catch ValueError, but that's likely in order to check if the signal is actually supported by the platform, and I don't think it's to suppress an error that occurred because of the thread we're in.

yes I already dig to here...

---

try to find a way fix but I am not sure its better

```diff

-            try:
+            if threading.current_thread() is threading.main_thread():
                 signal.signal(signal.SIGWINCH, self.old_sigwinch)
-            except ValueError:
-                # This can happen when called from a non-main thread
-                # (e.g., asyncio REPL). In this case, we skip signal restoration
-                # to avoid the "signal only works in main thread" error.
-                pass
+            # If we're not in the main thread, we simply skip signal restoration
+            # as signal.signal() would raise ValueError anyway
             del self.old_sigwinch
```

---

> When defining self.old_sigwinch, we try-catch ValueError, but that's likely in order to check if the signal is actually supported by the platform, and I don't think it's to suppress an error that occurred because of the thread we're in.

yes, only happens here can be value error.
so I think both way is ok I prefer the try catch way with comments way

the threading way need to understand why we use it here.
what do you think

---

After digging a bit more I think there is a way to correctly make this work:

> When threads are enabled, this function can only be called from [the main thread of the main interpreter](https://docs.python.org/3/library/signal.html#signals-and-threads); attempting to call it from other threads will cause a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError) exception to be raised.

> On Windows, [signal()](https://docs.python.org/3/library/signal.html#module-signal) can only be called with [SIGABRT](https://docs.python.org/3/library/signal.html#signal.SIGABRT), [SIGFPE](https://docs.python.org/3/library/signal.html#signal.SIGFPE), [SIGILL](https://docs.python.org/3/library/signal.html#signal.SIGILL), [SIGINT](https://docs.python.org/3/library/signal.html#signal.SIGINT), [SIGSEGV](https://docs.python.org/3/library/signal.html#signal.SIGSEGV), [SIGTERM](https://docs.python.org/3/library/signal.html#signal.SIGTERM), or [SIGBREAK](https://docs.python.org/3/library/signal.html#signal.SIGBREAK). A [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError) will be raised in any other case

So another way to be sure about what we're doing is:

- check if you're not on Windows (`os.name != 'nt'`)
- if this is the case, just silence the ValueError.

Rationale:

- When setting `old_sigwinch`, we suppress ValueError. This may happen if (1) we're on WIndows (2) we're not in the main thread
- When restoring the handler, we simply need to check we're not on windows. It shouldn't be possible to be in a non-main thread otherwise. But to be sure we can do the following:

```py
if os.name != 'nt':
    try:
        signal.signal(signal.SIGWINCH, self.old_sigwinch)
    except ValueError as e:
        # We can silence the ValueError if signal.signal() raised it
        # from a non-main thread on a non-Windows platform. Otherwise,
        # we need to re-raise as its cause could be different.
        if threading.current_thread() is threading.main_thread():
            raise e
```

---

learn it and will open every doc to learn more, thank you for the info

---

its perfect and I will change the code and add author

---

fixed and add author thank you again